packages/react-components/react-combobox/src/hooks/useComboboxFilter.tsx

```diff
@@ -0,0 +1,72 @@
+import * as React from 'react';
+import { Option } from '../Option';
+
+type UseComboboxFilterConfig<T extends { children: React.ReactNode; value: string } | string> = {
+  /** Provides a custom filter for the option. */
+  filter?: (optionText: string, query: string) => boolean;
+
+  /** Provides a custom message to display when there are no options. */
+  noOptionsMessage?: React.ReactNode;
+
+  /** Provides a way to map an option object to a React key. By default, "value" is used. */
+  optionToReactKey?: (option: T) => string;
+
+  /** Provides a way to map an option object to a text used for search. By default, "value" is used. */
+  optionToText?: (option: T) => string;
+
+  /** Provides a custom render for the option. */
+  renderOption?: (option: T) => JSX.Element;
+};
+
+function defaultFilter(optionText: string, query: string) {
+  if (query === '') {
+    return true;
+  }
+
+  return optionText.toLowerCase().includes(query.toLowerCase());
+}
+
+function defaultToString(option: string | { value: string }) {
+  return typeof option === 'string' ? option : option.value;
+}
+
+export function useComboboxFilter<T extends { children: React.ReactNode; value: string } | string>(
+  query: string,
+  options: T[],
+  config: UseComboboxFilterConfig<T>,
+) {
+  const {
+    filter = defaultFilter,
+    noOptionsMessage = "We couldn't find any matches.",
+    optionToReactKey = defaultToString,
+    optionToText = defaultToString,
+
+    renderOption = (option: T) => {
+      if (typeof option === 'string') {
+        return <Option key={option}>{option}</Option>;
+      }
+
+      return (
+        <Option {...option} key={optionToReactKey(option)} text={optionToText(option)} value={option.value}>
+          {option.children}
+        </Option>
+      );
+    },
+  } = config;
+
+  const filteredOptions = React.useMemo(() => {
+    const searchValue = query.trim();
+
+    return options.filter(option => filter(optionToText(option), searchValue));
+  }, [options, optionToText, filter, query]);
+
+  if (filteredOptions.length === 0) {
+    return [
+      <Option key="no-results" text="">
```

Comment:
with this, you would add a selectable option. Instead there should be just a visible text that you cannot select and it should be announced using aria-live

Reply:
This would be changing the current proposed pattern [in the docs](https://react.fluentui.dev/?path=/docs/components-combobox--default#filtering). Are we OK with that?
![image](https://github.com/microsoft/fluentui/assets/20744592/c9015713-6bec-402b-831b-e55cf2226554)

Reply:
I think we really want to avoid aria-live here, since live regions conflict with user actions like typing. The chance of a live region reliably presenting a no-results message in a situation like this is very low.

To address the issue with the no-results Option presenting as if it were selectable, we could add the `aria-disabled` state to it. In addition to that, I think the text itself should make the information clear enough.

Reply:
![image](https://github.com/microsoft/fluentui/assets/14183168/c682819b-d0b9-4b71-bea1-27080890ff5c)

@smhigley it seems that the approach with `aria-disabled` does not work properly, at least for NVDA. Do you have other suggestions?

Reply:
@smhigley even focusing the list doesn't work reliable... and it conflict with typing the same way... :( 
Please see my enclosed video. 
Even the item get focus and it should be narrated, the typed character interrupted it immediately and then user cannot hear the message anyway.  

As well I am aware of aria-live non reliability.

It think there is no suitable solution to work in 100% for every scenario.  

So if you like to keep focusing the "No found" message  I am fine with it. As there is not good way how to solve it. :(
https://github.com/microsoft/fluentui/assets/22620150/ad01b03f-4329-4e66-bf3d-db6e3f821307